File: compilers/python/tests/test_r237_dce_warning.py
"""R-237 (CL-GAP-013): the DCE warning for a dropped readonly field never
reached this tier's CLI.

The check exists — `_warn_dropped_readonly_fields`, called from
`_compile_from_source_str_with_result` — but `compile_from_source_collecting_warnings`,
which is what `__main__` calls, returned only the VALIDATOR's warnings. So the
notice was computed and thrown away at the boundary, and an author whose field
vanished from the locking script heard it from ts, go, zig and java but not from
here. Same shape as R-162 in the Go tier.
"""

import tempfile
from pathlib import Path

from runar_compiler.compiler import compile_from_source_collecting_warnings

UNREAD = """
import { SmartContract, assert } from 'runar-lang';

export class UnreadField extends SmartContract {
  readonly unused: bigint;
  readonly limit: bigint;

  constructor(unused: bigint, limit: bigint) {
    super(unused, limit);
    this.unused = unused;
    this.limit = limit;
  }

  public unlock(x: bigint): void {
    assert(x < this.limit);
  }
}
"""

READ = UNREAD.replace("assert(x < this.limit);", "assert(x < this.limit + this.unused);")


def _warnings_for(source: str) -> list[str]:
    with tempfile.TemporaryDirectory() as d:
        p = Path(d) / "UnreadField.runar.ts"
        p.write_text(source)
        _artifact, warnings = compile_from_source_collecting_warnings(str(p))
        return list(warnings)


def test_warns_when_a_readonly_field_is_dropped():
    warnings = _warnings_for(UNREAD)
    assert any("readonly field 'unused'" in w and "eliminated by DCE" in w for w in warnings), (
        f"the dropped-field notice did not reach the caller: {warnings!r}"
    )


def test_does_not_warn_when_the_field_is_read():
    warnings = _warnings_for(READ)
    assert not any("readonly field 'unused'" in w for w in warnings), (
        f"a referenced field must not warn: {warnings!r}"
    )


def test_does_not_warn_for_a_field_that_is_used():
    warnings = _warnings_for(UNREAD)
    assert not any("readonly field 'limit'" in w for w in warnings), (
        f"'limit' is read by unlock and must not warn: {warnings!r}"
    )

File: compilers/ruby/test/frontend/test_r237_dce_warning.rb
# frozen_string_literal: true

require_relative "../test_helper"
require "tmpdir"
require "runar_compiler"

# R-237 (CL-GAP-013): this tier emitted no warning when DCE dropped an
# un-annotated readonly field.
#
# The finding names Ruby. Measured across all seven, three tiers were silent:
# rust, python and ruby. The other four (ts, go, zig, java) told the author that
# a field they declared had been eliminated from the locking script entirely.
#
# Ruby genuinely had no check at all — this adds it, ported from
# compilers/go/frontend/embed_always_dce.go. (Rust and Python already HAD the
# check and dropped it at the CLI boundary; those are fixed in their own tiers.)
class TestR237DceWarning < Minitest::Test
  UNREAD = <<~TS
    import { SmartContract, assert } from 'runar-lang';

    export class UnreadField extends SmartContract {
      readonly unused: bigint;
      readonly limit: bigint;

      constructor(unused: bigint, limit: bigint) {
        super(unused, limit);
        this.unused = unused;
        this.limit = limit;
      }

      public unlock(x: bigint): void {
        assert(x < this.limit);
      }
    }
  TS

  READ = UNREAD.sub("assert(x < this.limit);", "assert(x < this.limit + this.unused);")

  def compile_with_warnings(source)
    Dir.mktmpdir do |dir|
      path = File.join(dir, "UnreadField.runar.ts")
      File.write(path, source)
      _artifact, warnings = RunarCompiler.compile_from_source_collecting_warnings(path)
      warnings
    end
  end

  def test_warns_when_a_readonly_field_is_dropped
    warnings = compile_with_warnings(UNREAD)
    matching = warnings.grep(/readonly field 'unused'.*eliminated by DCE/)
    refute_empty matching,
                 "a readonly field no method reads is eliminated from the locking " \
                 "script; the author must hear about it. Got: #{warnings.inspect}"
  end

  def test_does_not_warn_when_the_field_is_read
    warnings = compile_with_warnings(READ)
    assert_empty warnings.grep(/readonly field 'unused'.*eliminated by DCE/),
                 "a field that IS referenced must not warn: #{warnings.inspect}"
  end

  def test_does_not_warn_for_the_field_that_is_used
    warnings = compile_with_warnings(UNREAD)
    assert_empty warnings.grep(/readonly field 'limit'/),
                 "'limit' is read by unlock and must not warn: #{warnings.inspect}"
  end
end
